### PR TITLE
Bump pyo3 from 0.17.1 to 0.19.2

### PR DESCRIPTION
**Pull request by V02460**
Bump pyo3 version to 19.2 to keep up with the latest release and benefit from improved compatibility with Python 3.12. Pythonize is bumped to v0.19.0 to match pyo3.

Also bumps the minimum abi3 of pyo3 from py37 to py38 as Python 3.7 is EOL.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: rust/src/push/evaluator.rs [hunk 1]
```diff
@@ -105,6 +105,17 @@ impl PushRuleEvaluator {
     /// Create a new `PushRuleEvaluator`. See struct docstring for details.
     #[allow(clippy::too_many_arguments)]
     #[new]
+    #[pyo3(signature = (
```
Comment (reivilibre) [inline rust/src/push/evaluator.rs:108]: any chance for an explanation on what this is doing or why it's needed?

Comment (V02460) [reply rust/src/push/evaluator.rs:108]: New Pyo3 versions require that an explicit signature is given if any but the last parameter type is an `Option`. See https://pyo3.rs/main/function/signature#trailing-optional-arguments.

Comment (V02460) [reply rust/src/push/evaluator.rs:108]: If not used, clippy fails with a [deprectation message](https://github.com/matrix-org/synapse/actions/runs/5943525031/job/16118845691#step:5:26) that reads:

```
 error: use of deprecated constant `pyo3::impl_::deprecations::REQUIRED_ARGUMENT_AFTER_OPTION`: required arguments after an `Option<_>` argument are ambiguous and being phased out
       = help: add a `#[pyo3(signature)]` annotation on this function to unambiguously specify the default values for all optional parameters
   --> rust/src/push/evaluator.rs:113:9
    |
113 |         notification_power_levels: BTreeMap<String, i64>,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D deprecated` implied by `-D warnings`
```

**Review by erikjohnston — APPROVED** [review]
Thanks!

**Comment by erikjohnston** [comment]
And sorry this got dropped

**Pull request merged**